types/hyperscript/index.d.ts

```diff
@@ -2,12 +2,17 @@
 // Project: https://github.com/dominictarr/hyperscript
 // Definitions by: Mike Linkovich <https://github.com/spacejack>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 2.2
 
 declare module 'hyperscript' {
 
 	interface HyperScript {
-		/** Creates an HTML element */
-		(tagName: string, ...args: any[]): HTMLElement;
+		/** Creates a typed HTMLElement */
+		<T extends keyof HTMLElementTagNameMap>(tagName: T, attrs?: Object, ...children: any[]): HTMLElementTagNameMap[T];
+		/** Creates a typed HTMLElement via coercion */
+		<T extends HTMLElement>(tagName: string, attrs?: Object, ...children: any[]): T;
```

Comment:
I think the 3rd overload is redundant. It lets you call `h` with no type arguments and just gives you `HTMLElement` in return. But the second overload also gives you this, since `T` will fall back to its constraint, `HTMLElement`, when it can't infer anything from arguments.